docs/spec_draft.md

```diff
@@ -1266,6 +1267,56 @@ More formally, `result[i0, ..., iR-1]` is equal to:
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.reduce
+
+### Semantics
+
+Applies a `computation` function to `inputs` and `init_values` along the
+`dimensions` and produces a `result` tensor.
+
+### Inputs
+
+| Name          | Type                                             |
+|---------------|--------------------------------------------------|
+| `inputs`      | variadic number of tensors of any supported type |
+| `init_values` | variadic number of tensors of any supported type |
+| `dimensions`  | 1-dimensional tensor constant of type `si64`     |
+| `computation` | `function`                                       |
+
+### Outputs
+
+| Name     | Type                         |
+|----------|------------------------------|
+| `result` | tensor of any supported type |
+
+### Constraints
+
+  * (C1) `inputs`, `init_values` and `result` have the same element type.
+  * (C2) `inputs` and `init_values` have the same type.
+  * (C3) `inputs` have N tensors where N >= 1.
+  * (C4) 0 $\lt$ `dimensions[d]` $\lt$ rank(`inputs[d]`) for all dimension `d`.
+  * (C5) `computation` inputs have the same type as `inputs`, and output has the
+  same type as `result`.
+  * (C6) `init_values` have to form an identity under the `computation`.
```

Comment:
I believe so. From what I understand, every reduce iteration takes previous value and does computation with current value. Since the first iteration does not have a previous value, the init_value should provide an identity value.

The example in XLA doc shows two ways of computing reduction of 1-D array [10, 11, 12, 13]:
```
f(10, f(11, f(12, f(init_value, 13)))
but there are also many other possibilities, e.g.
f(init_value, f(f(10, f(init_value, 11)), f(f(init_value, 12), f(init_value, 13))))
```

Here, if the reduction is a summation and init_value was not 0, and say 1, the result of the two variations will be different.

Reply:
You are right.
I was mostly concerned about verification of this particular constraint. It seems that we should write something like

It is implementation defined behavior if the `initial value` has a effect on the `computation`